File: algorithms/sorts/quickSort.js
const { swap } = require("./utils");

// algorithm quicksort(A, lo, hi) is
//     if lo < hi then
//         p := partition(A, lo, hi)
//         quicksort(A, lo, p - 1)
//         quicksort(A, p + 1, hi)

// algorithm partition(A, lo, hi) is
//     pivot := A[hi]
//     i := lo
//     for j := lo to hi do
//         if A[j] < pivot then
//             swap A[i] with A[j]
//             i := i + 1
//     swap A[i] with A[hi]
//     return i

const partition = (itemList, start, end)=>{

    let leftRegionIndex = start;
    let rightRegionIndex = end;

    const pivot = itemList[Math.floor(((start+end)/2.0))];

    while(true){

        while(itemList[leftRegionIndex] < pivot){
            leftRegionIndex++;
            console.log(`LEFT_INDEX: ${leftRegionIndex}`);
        }

        while(itemList[rightRegionIndex] > pivot){
            rightRegionIndex--;
            console.log(`RIGHT: ${rightRegionIndex}`);
        }

        if(leftRegionIndex >= rightRegionIndex)
            break;
            
        itemList = swap(itemList, leftRegionIndex, rightRegionIndex);

        leftRegionIndex++;
        rightRegionIndex--;
    }
    return rightRegionIndex;

}

const quickSort = (itemList=[], start, end) =>{
    if(start < end){
        const p = partition(itemList, start, end);
        quickSort(itemList, start, p);
        quickSort(itemList, p+1, end);
    }
}


let itemList = new Array(20);


for(let index=0; index < itemList.length; index++)
    itemList[index] = Math.random() * 1000;


console.log("BEFORE SORT:");
console.log(JSON.stringify(itemList));

quickSort( itemList, 0, itemList.length-1);

console.log("AFTER SORT:");
console.log( JSON.stringify( itemList ) );

File: algorithms/sorts/c/quicksort.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <stdbool.h>


void swap(int *array, int i, int j);
int partition(int *array, int start, int end);
void quicksort(int *array, int start, int end);

void swap(int *array, int i, int j){
    int temp = array[i];
    array[i] = array[j];
    array[j] = temp;
}

int partition(int *array, int start, int end){
    int i = start;
    int j = end;
    int pivot = floor(((start+end)/2.0));
    while(true){

        while(array[i] < array[pivot]){
            i++;
        }

        while(array[j] > array[pivot]){
            j--;
        }

        if(i >= j){
            return j;
        }

        swap(array, i, j);
    }
}

void quicksort(int *array, int start, int end){
    if(start < end){
        int pivot = partition(array, start, end);
        quicksort(array, start, pivot);
        quicksort(array, pivot+1, end);
    }
}

int main(int argc, char **argv){
    int counter = 0;
    int array[20];
    printf("BEFORE SORT:\n");
    for(;counter < 20; counter++){
        array[counter] = rand();
        printf("%d ", array[counter]);
    }
    printf("\n");
    quicksort(array, 0, 20-1);

    printf("AFTER SORT:\n");
    for(counter=0;counter < 20; counter++){
        printf("%d ", array[counter]);
    }
    printf("\n");
    return 0;
}